File: 14-cache/python-sdk/semantic_caching.py
"""
Cost Katana Cache: Semantic Caching (Python)

Advanced semantic similarity caching.

Run: python 18-cache/python-sdk/semantic_caching.py
"""

import os
import requests

GATEWAY_URL = "https://cost-katana-backend.store/api/gateway/v1/chat/completions"
API_KEY = os.getenv("COST_KATANA_API_KEY")

def semantic_caching_demo():
    """Demonstrate semantic caching"""
    
    print("🥷 Semantic Caching Demonstration\n")
    
    # Original query
    print("1️⃣ Original query...")
    r1 = requests.post(
        GATEWAY_URL,
        json={
            "model": "gpt-4",
            "messages": [{"role": "user", "content": "What are the benefits of cloud computing?"}]
        },
        headers={
            "Authorization": f"Bearer {API_KEY}",
            "X-Enable-Cache": "true",
            "X-Semantic-Threshold": "0.90"
        }
    )
    r1.raise_for_status()
    print(f"  Cache Status: {r1.headers.get('X-Cache-Status')}")
    
    # Semantically similar
    print("\n2️⃣ Semantically similar query...")
    r2 = requests.post(
        GATEWAY_URL,
        json={
            "model": "gpt-4",
            "messages": [{"role": "user", "content": "Tell me the advantages of using cloud services"}]
        },
        headers={
            "Authorization": f"Bearer {API_KEY}",
            "X-Enable-Cache": "true"
        }
    )
    r2.raise_for_status()
    print(f"  Cache Status: {r2.headers.get('X-Cache-Status')}")
    print(f"  Similarity Score: {r2.headers.get('X-Similarity-Score')}")
    print(f"  Cost Saved: ${r2.headers.get('X-Cost-Saved')}")
    
    print("\n💡 Semantic Caching Features:")
    print("  ✅ Recognizes paraphrased queries")
    print("  ✅ Handles different phrasings")
    print("  ✅ Context-aware matching")
    print("  ✅ 70-90% hit rate typical")

def main():
    if not API_KEY:
        print("❌ COST_KATANA_API_KEY required")
        return
    
    try:
        semantic_caching_demo()
        print("\n✅ Semantic caching demo complete!")
    except requests.exceptions.HTTPError as e:
        print(f"❌ Error: {e.response.json() if e.response else e}")

if __name__ == "__main__":
    main()

File: 14-cache/npm-package/semantic-caching.ts
/**
 * Cost Katana Cache: Semantic Caching
 * 
 * Advanced semantic similarity caching for AI requests.
 * 
 * Run: npx ts-node 18-cache/npm-package/semantic-caching.ts
 */

import CostKatana from 'cost-katana';

const client = new CostKatana({
  apiKey: process.env.COST_KATANA_API_KEY!
});

async function semanticCachingDemo() {
  console.log('🥷 Semantic Caching Demonstration\n');

  // Original query
  console.log('1️⃣ Original query...');
  const response1 = await client.chat.completions.create({
    model: 'gpt-4',
    messages: [
      { role: 'user', content: 'What are the benefits of cloud computing?' }
    ]
  }, {
    headers: {
      'X-Enable-Cache': 'true',
      'X-Semantic-Threshold': '0.90'
    }
  });

  console.log(`  Cache Status: ${response1.headers?.['x-cache-status']}`);
  console.log(`  Response: ${response1.choices[0].message.content?.substring(0, 100)}...`);

  // Semantically similar query (different wording)
  console.log('\n2️⃣ Semantically similar query...');
  const response2 = await client.chat.completions.create({
    model: 'gpt-4',
    messages: [
      { role: 'user', content: 'Tell me the advantages of using cloud services' }
    ]
  }, {
    headers: { 'X-Enable-Cache': 'true' }
  });

  console.log(`  Cache Status: ${response2.headers?.['x-cache-status']}`);
  console.log(`  Similarity Score: ${response2.headers?.['x-similarity-score']}`);
  console.log(`  Cost Saved: $${response2.headers?.['x-cost-saved']}`);
  console.log(`  Original Query: ${response2.headers?.['x-original-query']}`);

  // Different context (cache miss)
  console.log('\n3️⃣ Different context (cache miss)...');
  const response3 = await client.chat.completions.create({
    model: 'gpt-4',
    messages: [
      { role: 'system', content: 'You are a cloud expert' },
      { role: 'user', content: 'What are the benefits of cloud computing?' }
    ]
  }, {
    headers: { 'X-Enable-Cache': 'true' }
  });

  console.log(`  Cache Status: ${response3.headers?.['x-cache-status']}`);
  console.log('  Reason: Different system message creates different context');

  console.log('\n💡 Semantic Caching Features:');
  console.log('  ✅ Recognizes paraphrased queries');
  console.log('  ✅ Handles different phrasings');
  console.log('  ✅ Context-aware matching');
  console.log('  ✅ Model-specific caching');
  console.log('  ✅ User-scoped caching');
}

async function main() {
  if (!process.env.COST_KATANA_API_KEY) {
    console.error('❌ COST_KATANA_API_KEY required');
    process.exit(1);
  }

  try {
    await semanticCachingDemo();
    console.log('\n✅ Semantic caching demo complete!');
  } catch (error) {
    console.error('\n❌ Failed:', error);
    process.exit(1);
  }
}

if (require.main === module) {
  main();
}

export { semanticCachingDemo };
